gcore: Windows x64 build fails with AVX2 enabled

<!--
IMPORTANT: Do NOT use GitHub to post any questions or support requests!
           They will be closed immediately and ignored.

Questions should go to the gdal-dev mailing list at
https://lists.osgeo.org/mailman/listinfo/gdal-dev
or other support forums. GitHub issues are for bug
reports and suggestions for new features.

If you think there is an issue with coordinate order in GDAL 3.0, then it is
likely an intended behavior. See https://github.com/OSGeo/gdal/issues/1974
for more explanations.

The GDAL project is made of contributions from various individuals and
organizations, each with their own focus. The issue you are facing is
not necessarily in the priority list of those contributors and consequently
there is no guarantee that it will be addressed in a timely manner.
If this bug report or feature request is high-priority for you, and you
cannot address it yourself, we suggest engaging a GDAL developer or support
organisation and financially sponsoring a fix.

-->

## Problem
When the MSVC `/arch:AVX2` flag is enabled for a Windows x64 build with Visual Studio 2022, there is a build error in `overview.cpp`.

Here is the error output:
```
C:\code\gdal\gcore\overview.cpp(343,38): error C2676: binary '*': 'Tsquare' does not define this operator or a conversion to a type acceptable to the predefined operator [C:\code\gdal\build\gcore\gcore.vcxproj]
          with
          [
              Tsquare=__m256d
          ]
C:\code\gdal\gcore\overview.cpp(770,1): message : see reference to function template instantiation 'Tsquare SQUARE<__m256d,T>(T)' being compiled [C:\code\gdal\build\gcore\gcore.vcxproj]
          with
          [
              Tsquare=__m256d,
              T=__m256d
          ]
C:\code\gdal\gcore\overview.cpp(1354,1): message : see reference to function template instantiation 'int QuadraticMeanUInt16SSE2<T>(int,int,const T *&,T *)' being compiled [C:\code\gdal\build\gcore\gcore.vcxproj]
          with
          [
              T=GByte
          ]
C:\code\gdal\gcore\overview.cpp(1779,57): message : see reference to function template instantiation 'CPLErr GDALResampleChunk32R_AverageOrRMS_T<GByte,int,GDT_Byte>(double,double,double,double,const T *,const GByte *,int,int,int,int,int,int,int,int,GDALRasterBand *,void **,const char *,int,float,GDALColorTable *,bool)' being compiled [C:\code\gdal\build\gcore\gcore.vcxproj]
          with
          [
              T=GByte
          ]
```

The problem is that there is no `SQUARE` function specific to the __m256d type. This is the line of code at overview.cpp, line 770:
```
const auto firstLineLoDbl = SQUARE(_mm256_cvtepi32_pd(firstLineLo));
```
Since there is no `SQUARE` function with a __m256d type, it uses the template `SQUARE` function at line 341:
```
template <class T, class Tsquare = T> inline Tsquare SQUARE(T val)
{
    return static_cast<Tsquare>(val) * val;
}
```
This will attempt to find an operator `*` for the __m256d type which will fail because it does not exist. There needs to be a `SQUARE` implementation for the __m256d type using an AVX2 intrinsic function.

## Details

I was confused why the GDAL Windows CI build was not displaying this problem because it also uses Visual Studio 2022 and enables the MSVC `/arch:AVX2` MSVC compiler flag. It looks like it is a difference between x86 and x64 builds.

The block of code that has build errors is compiled if `__AVX2__` is defined.

[Microsoft documentation](https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170) states that the `__AVX2__` preprocessor macro should be set for both x86 and x64 if `/arch:AVX2` is specified:

> __AVX2__ Defined as 1 when the [/arch:AVX2](https://learn.microsoft.com/en-us/cpp/build/reference/arch-x86?view=msvc-170) or [/arch:AVX512](https://learn.microsoft.com/en-us/cpp/build/reference/arch-x86?view=msvc-170) compiler option is set and the compiler target is x86 or x64. Otherwise, undefined.

It looks like the `__AVX2__` preprocessor macro is *not* actually set for x86 builds, only for x64 builds. Since the GDAL Windows CI only does an x86 build, you don't see this problem.

I'm not sure if this problem is specific to MSVC for Visual Studio 2022 or if it existed in earlier versions as well.

## Steps to reproduce the problem

This shows both a working x86 and a failing x64 setup. Having both makes it easy to compare.

### Failing x64 Build

* Make root directory:
  `cd C:\`
  `mkdir code`
  `cd code`
* Get *sqlite3* and *proj* through vcpkg
  `git clone https://github.com/Microsoft/vcpkg.git`
  `cd vcpkg`
  `.\bootstrap-vcpkg.bat`
  `vcpkg install sqlite3:x64-windows`
  `vcpkg install proj:x64-windows`
* Clone and build GDAL
  `cd C:\code`
  `git clone https://github.com/OSGeo/gdal.git`
  `cd gdal`
  `mkdir build`
  `cd build`
  `cmake -A x64 -G "Visual Studio 17 2022" -DCMAKE_C_FLAGS_INIT=/arch:AVX2 -DCMAKE_CXX_FLAGS_INIT=/arch:AVX2 -DPROJ_LIBRARY_RELEASE=C:\code\vcpkg\installed\x64-windows\lib\proj.lib -DPROJ_INCLUDE_DIR=C:\code\vcpkg\installed\x64-windows\include ..`
  `cmake --build . --target gcore`

You should be able to see the build errors.

### Working x86 Build

* Make root directory:
  `cd C:\`
  `mkdir code`
  `cd code`
* Get *sqlite3* and *proj* through vcpkg
  `git clone https://github.com/Microsoft/vcpkg.git`
  `cd vcpkg`
  `.\bootstrap-vcpkg.bat`
  `vcpkg install sqlite3`
  `vcpkg install proj`
* Clone and build GDAL
  `cd C:\code`
  `git clone https://github.com/OSGeo/gdal.git`
  `cd gdal`
  `mkdir build`
  `cd build`
  `cmake -A Win32 -G "Visual Studio 17 2022" -DCMAKE_C_FLAGS_INIT=/arch:AVX2 -DCMAKE_CXX_FLAGS_INIT=/arch:AVX2 -DPROJ_LIBRARY_RELEASE=C:\code\vcpkg\installed\x86-windows\lib\proj.lib -DPROJ_INCLUDE_DIR=C:\code\vcpkg\installed\x86-windows\include ..`
  `cmake --build . --target gcore`

gcore compiles without any problems.

--- a/gcore/overview.cpp
+++ b/gcore/overview.cpp
@@ -661,6 +661,11 @@ inline __m128d SQUARE(__m128d x)
 
 #ifdef __AVX2__
 
+inline __m256d SQUARE(__m256d x)
+{
+    return _mm256_mul_pd(x, x);
+}
+
 inline __m256d FIXUP_LANES(__m256d x)
 {
     return _mm256_permute4x64_pd(x, _MM_SHUFFLE(3, 1, 2, 0));

